### ISSUE TITLE
RegisterCollection from assemblies with scope

### ISSUE THREAD

**Issue by mrchief** [simpleinjector/SimpleInjector#215]
Not sure how to do this with new API

**v2.x code**
``` c#
var scopedLifestyle = new ExecutionContextScopeLifestyle();
container.RegisterManyForOpenGeneric(typeof(IRequestHandler<,>), scopedLifestyle, assemblies);
```

`RegisterCollection` doesn't take scope and `Scope.CreateRegistration` doesn't take assemblies.

**Comment by mrchief** [comment]
Currently I'm doing this but not sure if this is the right way:

```
var container = new Container();
container.Options.DefaultScopedLifestyle = new ExecutionContextScopeLifestyle();
// do other stuff
container.RegisterManyForOpenGeneric(typeof(IRequestHandler<,>), assemblies);
// more registrations
// override when needed
container.RegisterDecorator<IService, ServiceImpl>(Lifestyle.Singleton);
```

**Comment by mrchief** [comment]
Ok turns out its lot more...

```
public class ScopedLifestyleSelectionBehavior : ILifestyleSelectionBehavior
{
    public Lifestyle SelectLifestyle(Type serviceType, Type implementationType)
    {
        return Lifestyle.Scoped;
    }
}

...

var container = new Container();
container.Options.DefaultScopedLifestyle = new WebRequestLifestyle();
container.Options.LifestyleSelectionBehavior = new ScopedLifestyleSelectionBehavior();

// do other stuff
container.RegisterManyForOpenGeneric(typeof(IRequestHandler<,>), assemblies);

// more registrations

// override when needed
container.RegisterDecorator<IService, ServiceImpl>(Lifestyle.Singleton);
```

**Comment by mrchief** [comment]
Doing like above, I get `[Torn Lifestyle]` errors. :(

**Comment by dotnetjunkie** [comment]
What are the torn lifestyle errors you get?

**Comment by dotnetjunkie** [comment]
With this v2 syntax:

``` c#
container.RegisterManyForOpenGeneric(typeof(IRequestHandler<,>), scopedLifestyle, assemblies);
```

The equivalent syntax in v3 is:

``` c#
container.Register(typeof(IRequestHandler<,>), assemblies, scopedLifestyle);
```

or 

``` c#
container.Register(typeof(IRequestHandler<,>), assemblies, Lifestyle.Scoped);
```

**Comment by mrchief** [comment]
Ha! Now that you've said it, I feel utterly stupid! Why didn't I think of that? :grin: 

Its the same one that you mention [here](http://simpleinjector.readthedocs.org/en/latest/tornlifestyle.html). 

And I think I kind of understand what its trying to warn me against, but I think I'm ok with what I have and although I'd like to have a clean slate, I'm not ready to refactor everything yet. 

Let me give an example, bear with me please (and its no coincidence if you find things a little too [familiar](https://cuttingedge.it/blogs/steven/pivot/entry.php?id=92) :smile:):

```cs
// oversimplifying things here for brevity
public class FindUserByEmailQuery : IQuery<User>
{
    public string Email { get; set; }
}

public class FindUserByPhoneNumberQuery : IQuery<User>
{
    public long E164Number { get; set; }
}

public class FindUserQueryHandler
    : IQueryHandler<FindUserByEmailQuery, User>,
    : IQueryHandler<FindUserByPhoneNumberQuery, User>
{
    private readonly NorthwindUnitOfWork db;
 
    public FindUserQueryHandler(NorthwindUnitOfWork db)
    {
        this.db = db;
    }
 
    public User Handle(FindUserByEmailQuery query)
    {
        return (
            from user in this.db.Users
            where user.Email.Equals(query.Email)
            select user)
            .ToArray();
    }

    public User Handle(FindUserByPhoneNumberQuery query)
    {
        return (
            from user in this.db.Users
            where user.E164Number.Equals(query.E164Number)
            select user)
            .ToArray();
    }
}
```

So I get a `[Torn Lifestyle]` error saying each handler will get its own separate instance, which is OK for me. (I can get you the exact message later if you need).

I suspect that if I broke `FindUserQueryHandler` into `FindUserByEmailQueryHandler` and `FindUserByPhoneQueryHandler`, the error/warning will go away but that is something I don't want to do. We have a large number of such classes where we clumped similar queries and some of them share private helper methods for various utility purposes. It'll be a huge effort to isolate all of them and then refactor to have the shared common logic move outside and still stay local to these handlers (since no one else needs to know about them).

**Comment by dotnetjunkie** [comment]
The `Register(Type, IEnumerable<Assembly>, Lifestyle)` method ensures that for any class is created exactly one `Registration` instance, which means that alone can't be the reason you are getting these torn lifestyle exceptions. Take a look at the following test:

``` c#
// Arrange
var container = new Container();

container.Options.DefaultScopedLifestyle = new LifetimeScopeLifestyle();

container.Register(typeof(IQueryHandler<,>), new[] { typeof(FindUserQueryHandler).Assembly },
    Lifestyle.Scoped);

// Act
container.Verify();

using (container.BeginLifetimeScope())
{
    var q1 = container.GetInstance<IQueryHandler<FindUserByEmailQuery, User>>();
    var q2 = container.GetInstance<IQueryHandler<FindUserByPhoneNumberQuery, User>>();

    // Assert
    Assert.IsTrue(object.ReferenceEquals(q1, q2));
}
```

While using Simple Injector v2.1.3, with the classes you defined, the call to `container.Verify()` succeeds and during the execution of a single scope, both query handler abstractions result in the same instance (hence: the `Assert.IsTrue` succeeds).

So assuming that you use the latest version of Simple Injector, there must be something else going on that causes these exceptions.

**Comment by mrchief** [comment]
I haven't tried with `Register` yet (will do tonight). The error I described was by using `RegisterCollection` as described in my previous [comment](https://github.com/simpleinjector/SimpleInjector/issues/215#issuecomment-206668039).

I also edited my previous comments to say `RegisterCollection`. I just realized I had `RegisterManyForOpenGeneric` there!

**Comment by mrchief** [comment]
The warning goes away with `Register`. 

There is one other case where `Register` still gives TornLifestyle error; i.e. if the same handler handles both sync and async queries like so:

```
public class SyncQuery : IQuery<ResponseA> { ... }

public class AsyncQuery: IAsyncQuery<ResponseB> {..}

public class Handler : 
  IAsyncQueryHandler<AsyncQuery, ResponseB>,
  ISyncQueryHandler<SyncQuery, ResponseA> 
{
  public async Task<ResponseB> HandleAsync(AsyncQuery) { ... }

  public  ResponseA Handle (SyncQuery query) {...}
}
```

**Comment by dotnetjunkie** [comment]
Yes, that's correct. You have two different interfaces, which means that you have the following registrations:

``` c#
container.Register(typeof(ISyncQueryHandler<,>), assemblies, Lifestyle.Scoped);
container.Register(typeof(IAsyncQueryHandler<,>), assemblies, Lifestyle.Scoped);
```

Each call to `Register` ensures that it only uses one `Registration` object per type, but you lose that guarantee when you make two separate calls to `Register`.

I've got two questions though: why do you have both synchronous and asynchronous queries in the same application? And why are your queries required to be scoped?

**Comment by mrchief** [comment]
Thanks for confirming that. 

> why do you have both synchronous and asynchronous queries in the same application?
Various reasons, mostly since this application dates to pre-async/await era. Most of the queries have been refactored to be async but some linger on.

> And why are your queries required to be scoped?
Mostly due to async/await and `Task` based runs where the `DbContext` needs to be preserved after a web request is done.

Now this is something related but I'm not sure if I should raise it as a separate question. With all these changes with respect to v3, I'm running into scope issues within the MVC app. WebApi app seems fine. 

Here's an example:

```cs
var container = new Container();
container.Options.DefaultScopedLifestyle = new ExecutionContextScopeLifestyle(); 
container.Options.LifestyleSelectionBehavior = new ScopedLifestyleSelectionBehavior();

container.Register(...);  // bunch of registrations
```

Somewhere in MVC:

```cs
public class BaseController : Controller
{        
    protected IMediator Mediator => DependencyResolver.Current.GetService<IMediator>();
    
    protected async Task<T> RequestAsync<T>(IAsyncRequest<T> request)
    {
        var response = await Mediator.RequestAsync(request);
        if (response.HasException())
        {
            // simply rethrowing doesn't show the stacktrace of the original exception as YSOD
            throw new ApplicationException("Database query failed", response.Exception);
        }
        return response.Data;
    }
}

public class UserController : BaseController
{
    public async Task<ActionResult> UserProfile()
    {
        var response = await RequestAsync(new GetUserProfileQuery {UserId = User.Id});

        return View(response);
    }
}
```

Side note: The property in `BaseController` is an anti-pattern and I know it violates many things. However, I believe it was done to keep the boilerplate code (of making the query and checking for response/exception) out of every action method. Despite me cringing at it every time I look at it, I haven't been able to find a better way out.

Anyway, so this code worked in v2 without complaints and probably is the reason why I need to specify the scope in the first place. 

But v3 complains about scope mismatches. I've tried Hybrid Lifestyle but issue remains. I read the article about how Web API and MVC handle them differently and Web API is fine with above design (same BaseController black magic exists there as well and the registration is done the same way).

Is there something I'm missing with v3? I would hate to go back to v2.

**Comment by dotnetjunkie** [comment]
>> And why are your queries required to be scoped?

> Mostly due to async/await and Task based runs where the DbContext needs to be preserved after a web request is done and the black art shown below.

I don't understand that. Your query handlers depend on `DbContext`. Your `DbContext` needs to be scoped, but that doesn't mean that your query handlers need to be scoped. They can be both transient and scoped

> But v3 complains about scope mismatches.

Can you show the exact message you get?

**Comment by dotnetjunkie** [comment]
But I think Simple Injector does miss some features right now to make it possible/easy to batch-register types that have multiple interfaces with a non-transient lifestyle. Up until recently we mostly promoted the use of transient registrations, but I myself am promoting the use of singleton object graphs right now. When you do this, you get into the same kinds of trouble as you are right now with the scoped registrations; the support right now isn't great when you have multiple interfaces on the same type.

This is something that really needs to be improved, but we need some time to get this right. We want to make this as easy as possible, especially since I feel that singleton object graphs are actually the way to go. But we have to be careful not to introduce (too severe) breaking changes that affects too many people.

One simple addition for instance would be the introduction of a `Register(Type, IEnumerable<Registration>)` overload. That would allow you to do the following:

``` c#
var sq = container.GetTypesToRegister(typeof(ISyncQueryHandler<,>), assemblies);
var aq = container.GetTypesToRegister(typeof(IAsyncQueryHandler<,>), assemblies);

var registrations = (
	from type in sq.Concat(aq)
	let registration = Lifestyle.Scoped.CreateRegistration(type, container)
	select new { registration, sync = sq.Contains(type), async = aq.Contains(type) })
	.ToArray();
	
container.Register(typeof(ISyncQueryHandler<,>),
	from r in registrations
	where r.sync
	select r.registration);
	
container.Register(typeof(IAsyncQueryHandler<,>),
	from r in registrations
	where r.async
	select r.registration);
```

But we might even do better, and supply a `Register(IEnumerable<Type> serviceTypes, IEnumerable<Type> implementationTypes, Lifestyle)` overload. This would allow you to do the following:

``` c#
container.Register(new[] {
    typeof(ISyncQueryHandler<,>),
    typeof(IAsyncQueryHandler<,>) },
    assemblies,
    Lifestyle.Scoped);
```

Or we might even be able to allow batch-registered types to automatically use known registrations. That would allow you to simply do this:

``` c#
container.Register(typeof(ISyncQueryHandler<,>), assemblies, Lifestyle.Scoped);
container.Register(typeof(IAsyncQueryHandler<,>), assemblies, Lifestyle.Scoped);
```

But this is currently just hypothethical. Currently it's extremely hard to do this by hand, so when it comes to the torn lifestyle exception, I think you currently have two options:

- Either you suppress the torn lifestyle warning on (all) registrations.
- Or you register those types as transient.

**Comment by mrchief** [comment]
> but that doesn't mean that your query handlers need to be scoped

Not anymore. That issue went away with using `Register`. 

> Can you show the exact message you get?

Here they are:

with this configuration:

```cs
var container = new Container();

// no default scope, no selection behavior

// or with both below uncommented,
//container.Options.DefaultScopedLifestyle = new ExecutionContextScopeLifestyle();
//container.Options.LifestyleSelectionBehavior = new ScopedLifestyleSelectionBehavior();
```

I get 

> The configuration is invalid. The following diagnostic warnings were reported:
> -[Disposable Transient Component] DBContext is registered as transient, but implements IDisposable.
> -[Disposable Transient Component] UnitOfWork is registered as transient, but implements IDisposable.

As per the this [solution](http://simpleinjector.readthedocs.org/en/latest/disposabletransientcomponent.html), I added scope to it to make this warning go away:

```cs
container.Options.DefaultScopedLifestyle = new ExecutionContextScopeLifestyle();
container.Options.LifestyleSelectionBehavior = new ScopedLifestyleSelectionBehavior();
```

So that led me to this issue: 

> The IMediator is registered as 'Execution Context Scope' lifestyle, but the instance is requested outside the context of a Execution Context Scope.

when this line of code is hit (see [comment ](https://github.com/simpleinjector/SimpleInjector/issues/215#issuecomment-207445552) for detailed code):

```cs
protected IMediator Mediator => DependencyResolver.Current.GetService<IMediator>();
```

**Comment by dotnetjunkie** [comment]
Can you show the code for your mediator implementation? In my opinion your mediator should be a singleton.

**Comment by dotnetjunkie** [comment]
> Side note: The property in BaseController is an anti-pattern and I know it violates many things. However, I believe it was done to keep the boilerplate code (of making the query and checking for response/exception) out of every action method. Despite me cringing at it every time I look at it, I haven't been able to find a better way out.

IMO you should remove the base class completely. Base classes are evil, they lead to lots of trouble and as you already noted, calling `ServiceLocator.GetService` right from within the `BaseController`'s constructor "violates many things".

Instead I'd suggest defining your `UserController` as follows:

``` c#
public class UserController : Controller
{
    private readonly IMediator mediator;
	
    public UserController(IMediator mediator) {
        this.mediator = mediator;
    }	
    public async Task<ActionResult> UserProfile() {
        var response = await mediator.RequestAsync(new GetUserProfileQuery {UserId = User.Id });
        return View(response);
    }
}
```

And the exception behavior that you added to the base class can be moved to a decorator of `IMediator`:

``` c#
public class ExceptionMediatorDecorator : IMediator {
    private readonly IMediator mediator;
    public ExceptionMediatorDecorator(IMediator mediator) {
        this.mediator = mediator;
    }
    public async Task<T> RequestAsync<T>(IAsyncRequest<T> request) {
        var response = await mediator.RequestAsync(request);
        if (response.HasException()) {
            // simply rethrowing doesn't show the stacktrace of the original exception as YSOD
            throw new ApplicationException("Database query failed", response.Exception);
        }
        return response.Data;	
    }
}
```

You can register your mediator and decorator as follows:

``` c#
container.RegisterSingleton<IMediator>(
    new ExceptionMediatorDecorator(SimpleInjectorMediator(container)));
```

If removing the base controller is a refactoring that is too big to do at once, you can start of by removing the `ServiceLocator` call by changing the `BaseController` to the following:

``` c#
public class BaseController : Controller
{
	public IMediator Mediator { get; set; }
}
```

Now the `Mediator` property is a public writable property and we can configure Simple Injector to inject that property for you:

``` c#
container.RegisterInitializer<BaseController>(c => {
	c.Mediator = container.GetInstance<IMediator>();
});
```

This is far from ideal, but at least moves you away from calling back into the service locator from within the base class.

**Comment by mrchief** [comment]
Thanks Steven! 

I noticed `RegisterInitializer` and I'm using it to do just what you described. The decorator is a neat idea too and I'll ask the team to slowly start refactoring to use it. Gonna make the controller a lot easier to test! 

Also, I was able to get past all my scoping/lifestyle issues. Here's what I have finally:

```cs
var container = new Container();

// bunch of other registrations

...

// using Register instead of RegisterCollection
container.Register(typeof(IRequestHandler<,>), assemblies);
container.Register(typeof(IAsyncRequestHandler<,>), assemblies);


var registration = container.GetRegistration(typeof(DbContext)).Registration;
registration.SuppressDiagnosticWarning(DiagnosticType.DisposableTransientComponent, "Disposed by ASP.Net Framework");

registration = container.GetRegistration(typeof(IUnitOfWork)).Registration;
registration.SuppressDiagnosticWarning(DiagnosticType.DisposableTransientComponent, "Disposed by ASP.Net Framework");
```

So far, regression-wise, the app seems to be fine. Also, turns out that I don't have to do any scoping, just that suppression of warning (which seems to be OK as per the docs). I'll have to do some profiling to ensure all DbContext etc are getting disposed properly but so far, so good.

**Comment by mrchief** [comment]
> especially since I feel that singleton object graphs are actually the way to go.

Do you mean in web apps like MVC/WebAPI etc? Why do you feel so? ANd things like DbContext et. al. are better off disposed, aren't they?

**Comment by dotnetjunkie** [comment]
Whether or not your `DbContext`s need to be scoped, depends a lot on how you structure your application. When working with query handlers, it might often be fine when each handler gets its own `DbContext` instance (thus transient), but once you start passing entities from class to class, or queryables from class to class, those classes often need to make use of the same `DbContext` instance. That will bring you directly to the discussion of [why you often need one DbContext per request](https://stackoverflow.com/questions/10585478/one-dbcontext-per-web-request-why/10588594#10588594). So be careful about having transient `DbContext`s. You might want to reuse them during the request.

Also note that a `DbContext` is not magically 'Disposed by ASP.Net Framework', as your code literal seems to imply. Nothing in the ASP.NET framework will automatically go look for DbContext instances that go out of scope, and Entity Framework doesn't register its instances into ASP.NET; or at least, not out of the box. Not disposing a `DbContext` is usually not a big deal though.

>> especially since I feel that singleton object graphs are actually the way to go.

> Do you mean in web apps like MVC/WebAPI etc? Why do you feel so? ANd things like DbContext et. al. are better off disposed, aren't they?

Some time ago I wrote a blog post about [the anti-pattern of injecting runtime data into your components](https://www.cuttingedge.it/blogs/steven/pivot/entry.php?id=99). The bottom line of that blog post is:

> Don't inject runtime data into components during construction; it causes ambiguity, complicates the composition root with an extra responsibility and makes it extraordinarily hard to verify the correctness of your DI configuration. My advice is to let runtime data flow through the method calls of constructed object graphs.

And you know what? A `DbContext` actually **IS** runtime data. It's a big state bag with entities that you change and Save. You shouldn't inject a `DbContext` into any class directly. Instead you should hide your `DbContext` behind an abstraction, such as an `IContextProvider`. This allows you to retrieve the `DbContext` at runtime (after the object graph is built) by calling `IContextProvider.Context` or something similar. The `IContextProvider.Context` implementation will during that call ensure that the right `DbContext` instance will be returned. You could still use Simple Injector's Scoped lifestyle to manage the lifestyle of the `DbContext`, but you don't have to. You can choose.

When you do this, and refrain from injecting other runtime data into components, and make your components immutable (which are all advised practices), you can make most of your object graph singleton. Making them singleton, seems to complicate things, but actually simplifies thing tremendously. There is a range of diagnostic warnings that Simple Injector checks on that just don't exist anymore when your graphs are singletons. I'm talking about Lifestyle Mismatches, Torn Lifestyles, Ambiguous Lifestyles, Disposable Transient Components. Not only that, it will become considerably easier to configure your application, and in some cases, you might not even really need Simple Injector (or any DI container) to do this for you.

There are some problem spots though and that are the parts where you integrate with a framework. Both MVC and Web API for instance, force your controllers to be transient. That's a necessary evil we'll have to deal with, but that doesn't mean that all other components should be transient as well. And do note that with ASP.NET Core, this is all going to change. In ASP.NET Core 1 Controllers can actually be singletons. They can be reused over threads.

**Comment by mrchief** [comment]
> You might want to reuse them during the request.

Yes and no. Yes because I agree that it makes sense to reuse them during a request. No because EF doesn't play nice all the time. Especially if you do mind materializing every query (or else you need to enable `MARS`)

But anyway, this is why I tried scoping them which led to Torn Lifestyle error which led me to start scoping everything. From your comments though, it seems that I should be scoping them and disabling Torn Lifestyle errors, correct?

> Also note that a DbContext is not magically 'Disposed by ASP.Net Framework', as your code literal seems to imply. 

Glad you caught that. Yes, I worded that way intentionally to pique your brain. You're right and 

> Nothing in the ASP.NET framework will automatically go look for DbContext instances that go out of scope, 

this confirms my observations as well.

> Instead you should hide your DbContext behind an abstraction, such as an IContextProvider. This allows you to retrieve the DbContext at runtime (after the object graph is built) by calling IContextProvider.Context or something similar. The IContextProvider.Context implementation will during that call ensure that the right DbContext instance will be returned.

Funny how I thought that would be an anti-pattern. I did consider that at one point of time. It was few years back though and I was just beginning to take my first steps in the DI world. Looking back at memory lane, I'm not sure what my reasoning was back then. 

Reading rest of your comment is also interesting. The more I think about it, the more I realize I can actually live with singleton graph. I'll certainly start embracing it more.

**Comment by dotnetjunkie** [comment]
It's still unclear to me why you need your query handlers to be scoped. They only need to be scoped either if they contain mutable state that should be reused when a query is executed multiple time during a single request (which is a very unlikely scenario, since components shouldn't have any mutable state and they are not very likely to be called multiple times in the same request), or when they are injected into a consumer that has a scoped lifetime. Do note that a query handler doesn't have to become scoped if one of its dependencies (such as the DbContext) is scoped. The relationship is the other way around; a scoped consumer forces its dependencies to become (at least) scoped as well.

> From your comments though, it seems that I should be scoping them and disabling Torn Lifestyle errors, correct?

No, I would rather say that you should make your make your query handlers transient (or ideally singleton, but that might cause a bit more refactoring, and you'll currently run into the same torn lifestyle issues).

> Funny how I thought that would be an anti-pattern ... I'm not sure what my reasoning was back then.

What I think you mean is that many people consider hiding a `DbContext` behind an `IUnitOfWork` abstraction bad practice, since the `DbContext` itself implements the unit of work pattern. The argument of those people is that as long as you expose `IQueryable<T>` from your unit of work, it’s impossible to completely abstract your O/RM, since `IQueryable<T>` is a leaky abstraction. This is a fair argument, but abstraction the O/RM completely has never been the issue for me. The main argument for me is that when we use a DbContext directly in our application, we violate both the Interface Segregation Principle (ISP) and Dependency Inversion Principle (DIP). We violate ISP, since `DbContext` is a big class containing many methods, while none of the classes in our system use all the methods on the `DbContext`. DIP is violated twice, since `DbContext` is a concrete class -and- it is defined by an external library. These two violations make it hard to intercept calls to it and add cross-cutting concerns. Once very useful cross-cutting concern is automatic filtering of entities based on the identity of the user, but there are more that you can think of.

Instead, we can comply with the SOLID principles by defining our own unit of work (or similar, whatever is suited for our application) abstraction that ideally contains just a single member. Such member would usually be something like `IQueryable<T> Set<T>()`. Although the use of `IQueryable<T>` is still a DIP violation, we are in a much better place, because it's now much easier to decorate and intercept our unit of work and start filtering on `IQueryable<T>`s. And it prevents us from having to inject runtime data into our components as well.

**Comment by mrchief** [comment]
> The relationship is the other way around; a scoped consumer forces its dependencies to become (at least) scoped as well.

You're correct. I went back and changed my initializer to be this:

```
var container = new Container();
container.Options.DefaultScopedLifestyle = new ExecutionContextScopeLifestyle();

container.Register<DbContext, DbContext>(Lifestyle.Scoped);
container.Register(typeof(IRequestHandler<,>), assemblies);

container.RegisterMvcControllers(Assembly.GetExecutingAssembly());
container.Verify();
DependencyResolver.SetResolver(new SimpleInjector.Integration.Web.Mvc.SimpleInjectorDependencyResolver(container));
```

And I don't get Torn Lifestyle warnings. :clap: 
I think I was confused with few different errors happening at one time or the other. 

However, I do get this now:

> The DbContext is registered as 'Execution Context Scope' lifestyle, but the instance is requested outside the context of a Execution Context Scope.


**Stack Trace**

```
[ActivationException: The DbContext is registered as 'Execution Context Scope' lifestyle, but the instance is requested outside the context of a Execution Context Scope.]
   SimpleInjector.Scope.GetScopelessInstance(ScopedRegistration`2 registration) +168
   SimpleInjector.Scope.GetInstance(ScopedRegistration`2 registration, Scope scope) +73
   SimpleInjector.Advanced.Internal.LazyScopedRegistration`2.GetInstance(Scope scope) +158
   DynamicInstanceProducer85.GetInstance(Object[] constants) +304
   SimpleInjector.Internals.ConstantsClosure`1.GetInstance() +15
   SimpleInjector.InstanceProducer.GetInstance() +233
   SimpleInjector.Container.GetInstanceForRootType(Type serviceType) +74
   SimpleInjector.Container.GetInstance(Type serviceType) +57
   SimpleInjector.Integration.Web.Mvc.SimpleInjectorDependencyResolver.GetService(Type serviceType) +63
   System.Web.Mvc.DefaultControllerActivator.Create(RequestContext requestContext, Type controllerType) +41

[InvalidOperationException: An error occurred when trying to create a controller of type 'Web.Controllers.LoginController'. Make sure that the controller has a parameterless public constructor.]
   System.Web.Mvc.DefaultControllerActivator.Create(RequestContext requestContext, Type controllerType) +178
   System.Web.Mvc.DefaultControllerFactory.GetControllerInstance(RequestContext requestContext, Type controllerType) +76
   System.Web.Mvc.DefaultControllerFactory.CreateController(RequestContext requestContext, String controllerName) +88
   System.Web.Mvc.MvcHandler.ProcessRequestInit(HttpContextBase httpContext, IController& controller, IControllerFactory& factory) +194
   System.Web.Mvc.MvcHandler.BeginProcessRequest(HttpContextBase httpContext, AsyncCallback callback, Object state) +50
   System.Web.Mvc.MvcHandler.BeginProcessRequest(HttpContext httpContext, AsyncCallback callback, Object state) +48
   System.Web.Mvc.MvcHandler.System.Web.IHttpAsyncHandler.BeginProcessRequest(HttpContext context, AsyncCallback cb, Object extraData) +16
   System.Web.CallHandlerExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute() +103
   System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously) +155
```

However, I cannot have a parameterless controller since I need to inject logger, user services (to validate password, handle forgot password workflow etc).

I can probably refactor them out but then, any controller that depends on a service that depends on a scoped instance throws the same error (instance is requested outside the context of a Execution Context Scope). 

Also, I really appreciate your patience on this matter. If you prefer, you can close the original issue since that has been resolved and we can move this discussion somewhere else.

**Comment by dotnetjunkie** [comment]
Forget about the "Make sure that the controller has a parameterless public constructor" exception message that MVC throws. This message is misleading. The real error is in the inner exception, that you already quoted.

The execution context scope lifestyle and lifetime scope lifestyle are scoped lifestyles that need to be started and stopped explicitly. They only work if you explicitly start them yourself. Exception to this rule is Web API, where the Web API DependencyResolver implicitly starts and ends a execution context scope for you. This is however not the case in a MVC application. Within an MVC application, you will either have to start an execution context scope explicitly, or use the `WebRequestLifestyle`. The `WebRequestLifestyle` is created especially for MVC and Web Form applications.

Long story short, change your configuration to the following:

``` c#
var container = new Container();
container.Options.DefaultScopedLifestyle = new WebRequestLifestyle();
```

**Comment by mrchief** [comment]
Will that work with async/await actions? From what I gather from the docs, async/await needs `ExecutionContextScopeLifestyle`. 


Also, did a quick try and got the same error (albeit from a different place under one the async calls).

There's a big stack but in short, they mostly look like this:

> [HttpException (0x80004005): Error executing child request for handler 'System.Web.Mvc.HttpHandlerUtil+ServerExecuteHttpHandlerAsyncWrapper'.]

**Comment by dotnetjunkie** [comment]
You can use WebRequestLifestyle for async stuff wothout any problem. The WebRequestLifestyle works by storing items in the current HttpContext and that context is available when doing async as well. So for MVC: Use WebRequestLifestyle.

**Comment by mrchief** [comment]
With WebRequestLifestyle, I get the above error:

> [HttpException (0x80004005): Error executing child request for handler 'System.Web.Mvc.HttpHandlerUtil+ServerExecuteHttpHandlerAsyncWrapper'.]

**Comment by mrchief** [comment]
Nevermind... Found this weird piece of code which was causing the issue:

```
Task.Run(async () => await RequestAsync(new GetOrganizationLogoQuery { UserId = LoggedInUserId })).GetAwaiter().GetResult()
```

So this was not flowing the execution scope leading to above error. 

Now digging up on google, I think this [SO post](http://stackoverflow.com/a/12743302/244353) might explain why it was done like that.

So looks like either I change my query (`GetOrganizationLogoQuery`) to be synchronous or figure out how to extend the execution scope to above `Task.Run(...)` block.

**Comment by dotnetjunkie** [comment]
You can wrap your query handlers in a decorator that starts an execution context scope. This would allow you to run the query completely in isolation.

**Comment by mrchief** [comment]
Yes, of course... :)

**Comment by dotnetjunkie** [comment]
You'll need to create a decorator with a `Func<T>` decoratee factory, as described [here](https://simpleinjector.readthedocs.org/en/latest/advanced.html#decorators-with-func-t-decoratee-factories) and have a hybrid scoped lifestyle that mixes the web request and ECS lifestyle as described  [here](https://simpleinjector.readthedocs.org/en/latest/lifetimes.html#hybrid).

**Comment by mrchief** [comment]
Thanks for the pointers! Yes, I have seen those and should be implementing them by today.

**Comment by mrchief** [comment]
Bit confused now... 

This is what I have:

```cs
public class GetOrganizationLogoQuery : IAsyncRequest<LogoDto> {...}
internal class GetOrganizationLogoQueryHandler : IAsyncRequestHandler<GetOrganizationLogoQuery, LogoDto> {
...
public async Task<LogoDto> HandleAsync(GetOrganizationLogoQuery request) {...}
...
}
```

Now using decorators, how do I make it sync? 

Without decorators, this is how it'd become sync:

```cs
public class GetOrganizationLogoQuery : IRequest<LogoDto> {...}
internal class GetOrganizationLogoQueryHandler : IRequestHandler<GetOrganizationLogoQuery, LogoDto> {
...
public LogoDto Handle(GetOrganizationLogoQuery request) {...}
...
}
```

I guess if I use Hybrid lifestyle, then the following (existing approach) would work fine too?

```cs
[ChildActionOnly]
public PartialViewResult Logo()
{
    var logo = Task.Run(() =>
    {
        return RequestAsync(new GetOrganizationLogoQuery { UserId = LoggedInUserId });
    })
    .GetAwaiter()
    .GetResult();
}
```

**Comment by dotnetjunkie** [comment]
> Now using decorators, how do I make it sync? 

You can't. By definition, a decorator doesn't change the contract; a decorator is a class that implements the same interface as the decorated class. Since the decorated class returns a `Task<TResult>`, so has the decorator. What you are looking for is an adapter:

``` c#
public class SynchronizingRequestHandlerAdapter<TQuery, TResult> :
    IRequestHandler<TQuery, TResult>
{
    private readonly IAsyncRequestHandler<TQuery, TResult> asyncHandler;
    public SynchronizingRequestHandlerAdapter(IAsyncRequestHandler<TQuery, TResult> asyncHandler) {
        this.asyncHandler = asyncHandler;
    }
    public TResult Handle(TQuery request) {
        return this.asyncHandler.Handle(request).GetAwaiter().GetResult();        
    }
}
```

You can register this adapter as follows:

``` c#
container.RegisterConditional(typeof(IRequestHandler<,>),
    typeof(SynchronizingRequestHandlerAdapter<,>),
    c => !c.Handled);
```

Because of this registration, the `SynchronizingRequestHandlerAdapter<,>` functions as fallback registration in case there is other `IRequestHandler<,>` registration. In other words, this allows your async implementations to become synchronous, just by depending on `IRequestHandler<,>`.

By the way, IMO it's a design error to have both `IAsyncRequest<TResult>` and `ISyncRequest<TResult>` abstractions. Whether or not an operation is synchronous or asynchronous is of no concern to the query object and the consumer of that query. Making the handler asynchronous in the future would cause changes to the query as well.

**Comment by mrchief** [comment]
Thanks for showing the light yet again! 

> By the way, IMO it's a design error to have both IAsyncRequest<TResult> and ISyncRequest<TResult> abstractions. Whether or not an operation is synchronous or asynchronous is of no concern to the query object and the consumer of that query.

I completely agree. I'm using Mediator pattern and libs like [Mediatr](https://github.com/jbogard/mediatr) and [Shortbus](https://github.com/mhinze/shortbus) seem to have both - I'm assuming to better convey the type of usage (along with usual benefits of type safety, intellisense) - like if I're to just use `IRequest<Tin, Tout>` and I could both have async/sync handlers, it'd be hard to tell the usage. Of course, mixing async/sync itself is a bad idea but sometimes you're forced to compromise in existing projects where a lot of other variables decide what leeway you get.

I haven't given a lot of thought around it but I think there ought to be a better solution.

**Comment by dotnetjunkie** [comment]
> I'm using Mediator pattern and libs like Mediatr and Shortbus seem to have both

IMO you should not delegate the definition of your application's abstractions -and especially not the application's core abstractions- to a third party library. Your application should be in control over the abstractions, or as the [Dependency Inversion Principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle) describes it:

> the abstracts are owned by the upper/policy layers

So do learn from those libraries, since they incorperate those message based architectures in their core. Learn and copy, but do not depend on them for providing your with the abstractions, because you and your application should be in control over the abstractions.

**Comment by mrchief** [comment]
The adapter works like a charm. I got rid of `IRequest/IAsyncRequest` and its now just `IRequest`. 

Regarding your other points, I totally agree. I'll have to slowly refactor out this mediator (its not 3rd party library but is based heavily on those) since its basically shaking the foundation but I've felt (what you're describing) the same way many times. 

Thanks for all the help!! Greatly appreciated!
